ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/BlockAttestationsPool.java

```diff
@@ -0,0 +1,153 @@
+      // We only care about aggregated attestations
+      return;
+    }
+
+    Bytes32 attestationDataHash = newAttestation.getData().hash_tree_root();
+    AtomicBoolean oldBitlistPresent = new AtomicBoolean(true);
+    final Bitlist oldBitlist =
+        unprocessedAttestationsBitlist.computeIfAbsent(
+            attestationDataHash,
+            (key) -> {
+              oldBitlistPresent.set(false);
+              return newAttestation.getAggregation_bits().copy();
+            });
+
+    if (oldBitlistPresent.get()) {
+      if (!setBitsForNewAttestation(oldBitlist, newAttestation)) {
+        return;
+      }
+    }
+
+    ChainStorage.add(newAttestation, aggregateAttesationsQueue);
+
+    ChainStorage.add(attestationDataHash, newAttestation.getData().getSlot(), dataRootToSlot);
+  }
+
+  public void addAttestationProcessedInBlock(Attestation attestation) {
+    Bytes32 attestationDataHash = attestation.getData().hash_tree_root();
+    final Bitlist bitlist = processedAttestationsBitlist.computeIfAbsent(attestationDataHash -> {
+      ChainStorage.add(
+              attestationDataHash,
+              attestation.getData().getSlot(),
+              dataRootToSlot);
+      return attestation.getAggregation_bits().copy();
+    });
+
+    for (int i = 0; i < attestation.getAggregation_bits().getCurrentSize(); i++) {
+      if (attestation.getAggregation_bits().getBit(i) == 1) bitlist.setBit(i);
+    }
+  }
+
+  public SSZList<Attestation> getAggregatedAttestationsForBlockAtSlot(UnsignedLong slot) {
+    SSZList<Attestation> attestations =
+        new SSZList<>(Attestation.class, Constants.MAX_ATTESTATIONS);
+    int numAttestations = 0;
+
+    while (aggregateAttesationsQueue.peek() != null
+        && aggregateAttesationsQueue.peek().getData().getSlot().compareTo(slot) <= 0
+        && numAttestations < Constants.MAX_ATTESTATIONS) {
+
+      Attestation aggregate = aggregateAttesationsQueue.remove();
+      Bytes32 attestationHashTreeRoot = aggregate.getData().hash_tree_root();
+
+      // Check if attestation has already been processed in successful block
+      if (ChainStorage.get(attestationHashTreeRoot, this.processedAttestationsBitlist)
+          .isPresent()) {
+        Bitlist bitlist =
+            ChainStorage.get(
+                    aggregate.getData().hash_tree_root(), this.processedAttestationsBitlist)
+                .get();
+        List<Integer> oldAttesters = getAttesterIndicesIntoCommittee(bitlist);
+        List<Integer> newAttesters =
+            getAttesterIndicesIntoCommittee(aggregate.getAggregation_bits());
+        if (oldAttesters.containsAll(newAttesters)) {
+          continue;
+        }
+      }
+      attestations.add(aggregate);
+      numAttestations++;
+    }
+    return attestations;
+  }
+
+  /* // Memory Cleaning
+
+  public void cleanAttestationsUntilEpoch(UnsignedLong epoch) {
+    Iterator<Bytes32> it = processedAttestationsBitlist.keySet().iterator();
```

Comment:
I think you also need to drop attestations from `aggregateAttesationsQueue` as well?